From eb8c258dda3ffd73eb2b84d237ae301d07924296 Mon Sep 17 00:00:00 2001
From: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Date: Wed, 23 Jul 2014 13:27:01 -0400
Subject: [input] i8042: add firmware_id support

Message-id: <1406122030-5632-6-git-send-email-benjamin.tissoires@redhat.com>
Patchwork-id: 86345
O-Subject: [RHEL7.1 / RHEL7.0.z PATCH 05/14] Input: i8042 - add firmware_id support
Bugzilla: 1093449
RH-Acked-by: Aristeu S. Rozanski F. <arozansk@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Peter Hutterer <peter.hutterer@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1093449
Upstream: since v3.15, backported in stable
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7722437
Test: tested by myself on a t440s and a t410, was extensively tested in RHEL 6.6

commit a7c5868c3482127cb308c779b8a6460a3353c17f
Author: Hans de Goede <hdegoede@redhat.com>
Date:   Sat Apr 19 20:47:35 2014 -0700

    Input: i8042 - add firmware_id support

    Fill in the new serio firmware_id sysfs attribute for pnp instantiated
    8042 serio ports.

    Signed-off-by: Hans de Goede <hdegoede@redhat.com>
    Acked-by: Peter Hutterer <peter.hutterer@who-t.net>
    Signed-off-by: Dmitry Torokhov <dmitry.torokhov@gmail.com>

Signed-off-by: Benjamin Tissoires <benjamin.tissoires@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/input/serio/i8042-x86ia64io.h b/drivers/input/serio/i8042-x86ia64io.h
index 0ec9abb..381b20d 100644
--- a/drivers/input/serio/i8042-x86ia64io.h
+++ b/drivers/input/serio/i8042-x86ia64io.h
@@ -702,6 +702,17 @@ static int i8042_pnp_aux_irq;
 static char i8042_pnp_kbd_name[32];
 static char i8042_pnp_aux_name[32];
 
+static void i8042_pnp_id_to_string(struct pnp_id *id, char *dst, int dst_size)
+{
+ strlcpy(dst, "PNP:", dst_size);
+
+ while (id) {
+  strlcat(dst, " ", dst_size);
+  strlcat(dst, id->id, dst_size);
+  id = id->next;
+ }
+}
+
 static int i8042_pnp_kbd_probe(struct pnp_dev *dev, const struct pnp_device_id *did)
 {
  if (pnp_port_valid(dev, 0) && pnp_port_len(dev, 0) == 1)
@@ -718,6 +729,8 @@ static int i8042_pnp_kbd_probe(struct pnp_dev *dev, const struct pnp_device_id *
   strlcat(i8042_pnp_kbd_name, ":", sizeof(i8042_pnp_kbd_name));
   strlcat(i8042_pnp_kbd_name, pnp_dev_name(dev), sizeof(i8042_pnp_kbd_name));
  }
+ i8042_pnp_id_to_string(dev->id, i8042_kbd_firmware_id,
+          sizeof(i8042_kbd_firmware_id));
 
  /* Keyboard ports are always supposed to be wakeup-enabled */
  device_set_wakeup_enable(&dev->dev, true);
@@ -742,6 +755,8 @@ static int i8042_pnp_aux_probe(struct pnp_dev *dev, const struct pnp_device_id *
   strlcat(i8042_pnp_aux_name, ":", sizeof(i8042_pnp_aux_name));
   strlcat(i8042_pnp_aux_name, pnp_dev_name(dev), sizeof(i8042_pnp_aux_name));
  }
+ i8042_pnp_id_to_string(dev->id, i8042_aux_firmware_id,
+          sizeof(i8042_aux_firmware_id));
 
  i8042_pnp_aux_devices++;
  return 0;
diff --git a/drivers/input/serio/i8042.c b/drivers/input/serio/i8042.c
index 78e4de4..71153db 100644
--- a/drivers/input/serio/i8042.c
+++ b/drivers/input/serio/i8042.c
@@ -87,6 +87,8 @@ MODULE_PARM_DESC(debug, "Turn i8042 debugging mode on and off");
 #endif
 
 static bool i8042_bypass_aux_irq_test;
+static char i8042_kbd_firmware_id[128];
+static char i8042_aux_firmware_id[128];
 
 #include "i8042.h"
 
@@ -1213,6 +1215,8 @@ static int __init i8042_create_kbd_port(void)
  serio->dev.parent = &i8042_platform_device->dev;
  strlcpy(serio->name, "i8042 KBD port", sizeof(serio->name));
  strlcpy(serio->phys, I8042_KBD_PHYS_DESC, sizeof(serio->phys));
+ strlcpy(serio->firmware_id, i8042_kbd_firmware_id,
+  sizeof(serio->firmware_id));
 
  port->serio = serio;
  port->irq = I8042_KBD_IRQ;
@@ -1239,6 +1243,8 @@ static int __init i8042_create_aux_port(int idx)
  if (idx < 0) {
   strlcpy(serio->name, "i8042 AUX port", sizeof(serio->name));
   strlcpy(serio->phys, I8042_AUX_PHYS_DESC, sizeof(serio->phys));
+  strlcpy(serio->firmware_id, i8042_aux_firmware_id,
+   sizeof(serio->firmware_id));
   serio->close = i8042_port_close;
  } else {
   snprintf(serio->name, sizeof(serio->name), "i8042 AUX%d port", idx);
-- 
1.7.1